### PR TITLE
8255560: Class::isRecord should check that the current class is final and not abstract

### PR DESCRIPTION
**Pull request by ChrisHegarty**
Update Class::isRecord to only return true for classes that are final.

The removal of non-specified JVM checks on classes with a Record Attribute (see JDK-8255342), has resulted in more types of loadable classes that may contain a Record Attribute. Since these checks are not performed by the JVM anymore, Class::isRecord, and by extension Class::getRecordComponents, may return true or component values, respectively, for classes that are not well-formed record classes (as per the JLS), .e.g. non-final or abstract classes, that contain a record Attribute.

Core Reflection, Class::isRecord, already asserts checks that the JVM does not, e.g. that the direct superclass is java.lang.Record. Some points from the Java Language Specification for record classes:

 1. It is a compile-time error if a record declaration has the modifier abstract.
 2. A record declaration is implicitly final.
 3. The direct superclass type of a record class is Record.

Class::isRecord already ensures no.3. This issue proposes to add explicit checks in Core Reflection to ensure no.1 and no.2, since the JVM now allows such classes that contain a Record Attribute to be loaded.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255560](https://bugs.openjdk.java.net/browse/JDK-8255560): Class::isRecord should check that the current class is final and not abstract

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to d626eeed9a452e2e3cd7465726fb911f31420c76
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to 4eb264883a9e6105faf3fbb27a9d2caf7a19eecf

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1543/head:pull/1543`
`$ git checkout pull/1543`

### CODE REVIEW

File: test/jdk/java/lang/reflect/records/IsRecordTest.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        assertEquals(subFooCls.getRecordComponents(), null);
+    }
+
+    /** Tests record-ness properties of traditionally compiled classes. */
+    @Test
+    public void testBasicRecords() {
+        out.println("\n--- testBasicRecords ---");
+        record EmptyRecord () { }
+        assertTrue(EmptyRecord.class.isRecord());
+        assertTrue(EmptyRecord.class.getRecordComponents() != null);
```
Comment (mlchung) [inline test/jdk/java/lang/reflect/records/IsRecordTest.java:138]: Better to have a more precise check for empty array

File: src/java.base/share/classes/java/lang/Class.java [hunk 2]
```diff
@@ -3664,9 +3664,10 @@ public boolean isEnum() {
      * Returns {@code true} if and only if this class is a record class.
      *
      * <p> The {@linkplain #getSuperclass() direct superclass} of a record
-     * class is {@code java.lang.Record}. A record class has (possibly zero)
-     * record components, that is, {@link #getRecordComponents()} returns a
-     * non-null value.
+     * class is {@code java.lang.Record}. A record class is
+     * {@link Modifier#FINAL}. A record class has (possibly zero) record
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/Class.java:3668]: nit: `{@linkplain Modifier#FINAL final}`.

**Review by mlchung — APPROVED** [review]
The change looks good with a couple of minor suggestion.

File: src/java.base/share/classes/java/lang/Class.java (hunk 2 above)
Comment (ChrisHegarty) [reply src/java.base/share/classes/java/lang/Class.java]: Good idea - done.

File: test/jdk/java/lang/reflect/records/IsRecordTest.java (hunk 1 above)
Comment (ChrisHegarty) [reply test/jdk/java/lang/reflect/records/IsRecordTest.java]: Good idea - done.

**Review by mlchung — APPROVED** [review]
`{@link ... final}` should be `@linkplain`.  Otherwise, looks good.

**Comment by ChrisHegarty** [comment]
> `{@link ... final}` should be `@linkplain`. Otherwise, looks good.

Oops! Yes, changed.

**Review by jddarcy — APPROVED** [review]

**Pull request closed without merging**